File: app/src/main/java/me/tysheng/xishi/net/XishiRetrofit.java
package me.tysheng.xishi.net;

import android.content.Context;

import java.io.File;
import java.io.IOException;
import java.util.concurrent.TimeUnit;

import me.tysheng.xishi.utils.SystemUtil;
import okhttp3.Cache;
import okhttp3.CacheControl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.fastjson.FastJsonConverterFactory;

/**
 * Created by shengtianyang on 16/3/19.
 */
public class XishiRetrofit {

    private static String BASE_URL = "http://dili.bdatu.com/jiekou/";
    private Context mContext;
    private volatile Retrofit retrofit = null;
    private volatile XishiService sService = null;
    private Interceptor REWRITE_CACHE_CONTROL_INTERCEPTOR = new Interceptor() {
        @Override
        public Response intercept(Chain chain) throws IOException {
            Request request = chain.request();
            CacheControl cacheControl = SystemUtil.isNetworkAvailable(mContext) ?
                    CacheControl.FORCE_NETWORK : CacheControl.FORCE_CACHE;
            request = request.newBuilder()
                    .cacheControl(cacheControl)
                    .build();
            Response originalResponse = chain.proceed(request);
            return originalResponse.newBuilder()
                    .build();
        }
    };

    public XishiRetrofit(Context context) {
        mContext = context;
    }

    private void init() {
        final File baseDir = mContext.getCacheDir();
        Cache cache = null;
        if (baseDir != null) {
            final File cacheDir = new File(baseDir, "HttpCache");
            //设置缓存 10M
            cache = new Cache(cacheDir, 10 * 1024 * 1024);
        }

        OkHttpClient.Builder builder = new OkHttpClient.Builder();
        //设置超时
        final int TIME_MAX = 12;
        builder.readTimeout(TIME_MAX, TimeUnit.SECONDS);
        builder.connectTimeout(TIME_MAX, TimeUnit.SECONDS);
        builder.writeTimeout(TIME_MAX, TimeUnit.SECONDS);
        //错误重连
        builder.retryOnConnectionFailure(true);
        //网络拦截
        builder.addInterceptor(REWRITE_CACHE_CONTROL_INTERCEPTOR);
        //缓存设置
        builder.cache(cache);

        OkHttpClient client = builder.build();

        retrofit = new Retrofit.Builder()
                .client(client)
                .baseUrl(BASE_URL)
                .addConverterFactory(FastJsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();

    }

    public XishiService get() {
        init();
        sService = retrofit.create(XishiService.class);
        return sService;
    }

}

File: app/src/main/java/me/tysheng/xishi/net/XishiRetrofit.kt
package me.tysheng.xishi.net

import android.content.Context
import com.google.gson.FieldNamingPolicy
import com.google.gson.GsonBuilder
import io.reactivex.schedulers.Schedulers
import me.tysheng.xishi.utils.SystemUtil
import okhttp3.Cache
import okhttp3.CacheControl
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import timber.log.Timber
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * Created by shengtianyang on 16/3/19.
 */
class XishiRetrofit(private val mContext: Context) {
    private lateinit var retrofit: Retrofit
    private lateinit var ngService: NgService
    private lateinit var xishiService: XishiService
    private lateinit var okHttpClient: OkHttpClient
    private val REWRITE_CACHE_CONTROL_INTERCEPTOR = Interceptor { chain ->
        var request = chain.request()
        val cacheControl = if (SystemUtil.isNetworkAvailable(mContext))
            CacheControl.FORCE_NETWORK
        else
            CacheControl.FORCE_CACHE
        request = request.newBuilder()
                .cacheControl(cacheControl)
                .build()
        val originalResponse = chain.proceed(request)
        originalResponse.newBuilder()
                .build()
    }
    private val gson= GsonBuilder()
            .setFieldNamingStrategy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
            .setExclusionStrategies()
            .create()

    private val authInterceptor = Interceptor {
        val oldRequest = it.request()
        val builder = oldRequest.newBuilder()
        builder.header("Authorization", "Bearer " + XishiService.token)
        it.proceed(builder.build())
    }

    init {
        okHttpClient()
    }

    fun createNgService(): NgService {
        retrofit = Retrofit.Builder()
                .client(okHttpClient)
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createWithScheduler(Schedulers.io()))
                .build()
        ngService = retrofit.create<NgService>(NgService::class.java)
        return ngService

    }

    fun createXishiService(): XishiService {
        retrofit = Retrofit.Builder()
                .client(okHttpClient)
                .baseUrl(XISHI_BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createWithScheduler(Schedulers.io()))
                .build()
        xishiService = retrofit.create<XishiService>(XishiService::class.java)
        return xishiService
    }

    private fun okHttpClient(): OkHttpClient? {
        val baseDir = mContext.cacheDir
        var cache: Cache? = null
        if (baseDir != null) {
            val cacheDir = File(baseDir, "HttpCache")
            cache = Cache(cacheDir, (100 * 1024 * 1024).toLong())
        }

        val builder = OkHttpClient.Builder()
        builder.readTimeout(TIME_MAX.toLong(), TimeUnit.SECONDS)
        builder.connectTimeout(TIME_MAX.toLong(), TimeUnit.SECONDS)
        builder.writeTimeout(TIME_MAX.toLong(), TimeUnit.SECONDS)
        builder.retryOnConnectionFailure(true)
        //builder.addInterceptor(REWRITE_CACHE_CONTROL_INTERCEPTOR)
        builder.addInterceptor(authInterceptor)
        builder.addInterceptor(HttpLoggingInterceptor(
                HttpLoggingInterceptor.Logger { message -> Timber.tag("OkHttp").d(message) }).apply {
            level = HttpLoggingInterceptor.Level.BODY
        })
        builder.cache(cache)

        okHttpClient = builder.build()
        return okHttpClient
    }

    companion object {
        private const val TIME_MAX = 12

        private const val BASE_URL = "http://dili.bdatu.com/jiekou/"
        private const val XISHI_BASE_URL = "http://132.232.188.197:8080/"
    }

}

